functions/src/functions/users.ts

```diff
@@ -0,0 +1,340 @@
+          result[userId] = {
+            error: {
+              code: error.code,
+              message: error.message,
+            },
+          }
+        } else if (error instanceof Error) {
+          result[userId] = {
+            error: {
+              code: '500',
+              message: error.message,
+            },
+          }
+        } else {
+          result[userId] = {
+            error: {
+              code: '500',
+              message: 'Internal server error',
+            },
+          }
+        }
+      }
+    }
+    return result
+  },
+)
+
+export interface UpdateUserInformationInput {
+  userId?: string
+  data?: UserInformation
+}
+
+export const updateUserInformationFunction = onCall(
+  async (request: CallableRequest<UpdateUserInformationInput>) => {
+    if (!request.auth?.uid)
+      throw new https.HttpsError('unauthenticated', 'User is not authenticated')
+
+    if (!request.data.userId)
+      throw new https.HttpsError('invalid-argument', 'User ID is required')
+
+    if (!request.data.data)
+      throw new https.HttpsError('invalid-argument', 'User data is required')
+
+    const firestoreService = new FirestoreService()
+    const user = await firestoreService.getUser(request.data.userId)
+    const organization = user.content?.organization
+
+    const securityService = new SecurityService()
+    try {
+      await securityService.ensureClinician(request.auth, organization)
+    } catch {
+      securityService.ensureUser(request.auth, request.data.userId)
+    }
+
+    const auth = admin.auth()
+    await auth.updateUser(request.data.userId, {
+      displayName: request.data.data.auth.displayName,
+      email: request.data.data.auth.email,
+      phoneNumber: request.data.data.auth.phoneNumber,
+      photoURL: request.data.data.auth.photoURL,
+    })
+  },
+)
+
+export interface DeleteUserInput {
+  userId?: string
+}
+
+export const deleteUserFunction = onCall(
+  async (request: CallableRequest<DeleteUserInput>) => {
+    if (!request.auth?.uid)
+      throw new https.HttpsError('unauthenticated', 'User is not authenticated')
+
+    if (!request.data.userId)
+      throw new https.HttpsError('invalid-argument', 'User ID is required')
+
+    const firestoreService = new FirestoreService()
+    const user = await firestoreService.getUser(request.auth.uid)
+
+    const securityService = new SecurityService()
+    await securityService.ensureClinician(
+      request.auth,
+      user.content?.organization,
+    )
+
+    const firestore = admin.firestore()
+    await firestore.recursiveDelete(
+      firestore.doc(`admins/${request.data.userId}`),
+    )
+    await firestore.recursiveDelete(
+      firestore.doc(`clinicians/${request.data.userId}`),
+    )
+    await firestore.recursiveDelete(
+      firestore.doc(`patients/${request.data.userId}`),
+    )
+    await firestore.recursiveDelete(
+      firestore.doc(`users/${request.data.userId}`),
+    )
```

Comment:
If these operations can be run in parallel (I think so?), `Promise.all` can be used to run them at the same time

Reply:
Technically, yes, but I´m not sure about what happens if one of them fails in general. Unfortunately, it doesn't seem to be possible to put these operations in a transaction

Reply:
Good point @pauljohanneskraft; let's note this down as something that we should cover here. Would be interesting to hear if this works with transactions and if it does not, now we might be able to use an alternative. Maybe also out of scope for this PR for now but a good future ticket.